### PR TITLE
 [SR GeekCamp] Improve schema scan performance #4028

### PR DESCRIPTION
**Pull request by ylcq**
improve schema scan performance continue.

1. fix bug: push down limit parameter only when there is no predicate
2. for "select * from information_schema.columns limit 10;" query, we can call describe_table interface once and
   no longer call get_db_names and get_table_names interface, which can reduce RPC time from BE to FE, and the amount of data

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 0 / 58 (00.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|0|58|00.00%|

File: be/src/exec/vectorized/schema_scan_node.cpp [hunk 1]
```diff
@@ -64,8 +64,12 @@ Status SchemaScanNode::init(const TPlanNode& tnode, RuntimeState* state) {
         _scanner_param.thread_id = tnode.schema_scan_node.thread_id;
     }
 
-    if (tnode.limit > 0) {
-        _scanner_param.limit = tnode.limit;
+    // only for no predicate and limit parameter is set
+    if (tnode.conjuncts.empty()) {
```
Comment (kangkaisen) [inline be/src/exec/vectorized/schema_scan_node.cpp:68]:
```suggestion
    if (tnode.conjuncts.empty() && tnode.limit > 0) {
```

File: be/src/exec/vectorized/schema_scanner/schema_columns_scanner.cpp [hunk 2]
```diff
@@ -197,19 +200,31 @@ Status SchemaColumnsScanner::fill_chunk(ChunkPtr* chunk) {
             // TABLE_SCHEMA
             {
                 ColumnPtr column = (*chunk)->get_column_by_slot_id(2);
-                std::string db_name = SchemaHelper::extract_db_name(_db_result.dbs[_db_index - 1]);
-                Slice value(db_name.c_str(), db_name.length());
-                fill_column_with_slot<TYPE_VARCHAR>(column.get(), (void*)&value);
+                if (_param->without_db_table) {
+                    std::string db_name = _desc_result.columns[_column_index].columnDesc.dbName;
+                    Slice value(db_name.c_str(), db_name.length());
```
Comment (kangkaisen) [inline be/src/exec/vectorized/schema_scanner/schema_columns_scanner.cpp:205]: The following codes are redundant，could write only once
```
                  Slice value(db_name.c_str(), db_name.length());
                    fill_column_with_slot<TYPE_VARCHAR>(column.get(), (void*)&value);
```

File: be/src/exec/vectorized/schema_scan_node.cpp (hunk 1 above)
Comment (ylcq) [reply be/src/exec/vectorized/schema_scan_node.cpp]: has been modified

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 0 / 58 (00.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|0|58|00.00%|

File: fe/fe-core/src/main/java/com/starrocks/service/FrontendServiceImpl.java [hunk 3]
```diff
@@ -471,62 +472,116 @@ public TDescribeTableResult describeTable(TDescribeTableParams params) throws TE
...
+    // dbs and tables, when reach limit, we break;
+    private void describeWithoutDbAndTable(UserIdentity currentUser, List<TColumnDef> columns, long limit) {
+        Catalog catalog = Catalog.getCurrentCatalog();
+        List<String> dbNames = catalog.getDbNames();
+        boolean reachLimit;
+        for (String fullName : dbNames) {
+            if (!Catalog.getCurrentCatalog().getAuth().checkDbPriv(currentUser, fullName, PrivPredicate.SHOW)) {
+                continue;
+            }
+            String dbName = fullName.indexOf(":") >= 0 ? fullName.substring(fullName.indexOf(":") + 1) : fullName;
```
Comment (mofeiatwork) [inline fe/fe-core/src/main/java/com/starrocks/service/FrontendServiceImpl.java:512]: Not so robust?

File: fe/fe-core/src/main/java/com/starrocks/service/FrontendServiceImpl.java [hunk 4]
```diff
@@ -471,62 +472,116 @@ public TDescribeTableResult describeTable(TDescribeTableParams params) throws TE
...
+            Database db = Catalog.getCurrentCatalog().getDb(fullName);
+            if (db != null) {
```
Comment (mofeiatwork) [inline fe/fe-core/src/main/java/com/starrocks/service/FrontendServiceImpl.java:514]: Do we need to throw exception here?

File: fe/fe-core/src/main/java/com/starrocks/service/FrontendServiceImpl.java [hunk 5]
```diff
@@ -471,62 +472,116 @@ public TDescribeTableResult describeTable(TDescribeTableParams params) throws TE
...
+        }
+    }
+
+    private boolean setColumnDesc(List<TColumnDef> columns, Table table, long limit,
+                               boolean needSetDbAndTable, String db, String tbl) {
+        if (table != null) {
+            String tableKeysType = "";
+            if (TableType.OLAP.equals(table.getType())) {
+                OlapTable olapTable = (OlapTable) table;
+                tableKeysType = olapTable.getKeysType().name().substring(0, 3).toUpperCase();
```
Comment (mofeiatwork) [inline fe/fe-core/src/main/java/com/starrocks/service/FrontendServiceImpl.java:542]: What does it means by `substring(0, 3)`?

File: be/src/exec/vectorized/schema_scanner/schema_columns_scanner.cpp [hunk 6]
```diff
@@ -197,7 +200,12 @@ Status SchemaColumnsScanner::fill_chunk(ChunkPtr* chunk) {
             // TABLE_SCHEMA
             {
                 ColumnPtr column = (*chunk)->get_column_by_slot_id(2);
-                std::string db_name = SchemaHelper::extract_db_name(_db_result.dbs[_db_index - 1]);
+                std::string db_name;
+                if (_param->without_db_table) {
+                    db_name = _desc_result.columns[_column_index].columnDesc.dbName;
```
Comment (stdpain) [inline be/src/exec/vectorized/schema_scanner/schema_columns_scanner.cpp:205]: could we set the dbname once?

File: fe/fe-core/src/main/java/com/starrocks/service/FrontendServiceImpl.java (hunk 3 above)
Comment (ylcq) [reply fe/fe-core/src/main/java/com/starrocks/service/FrontendServiceImpl.java:512]: Do you mean "String dbName = fullName.indexOf(":") >= 0 ? fullName.substring(fullName.indexOf(":") + 1) : fullName;" ?
ok, we can can process it in be,  then reuse be processing logic,

File: fe/fe-core/src/main/java/com/starrocks/service/FrontendServiceImpl.java (hunk 4 above)
Comment (ylcq) [reply fe/fe-core/src/main/java/com/starrocks/service/FrontendServiceImpl.java:514]: This is the previous code logic, which has not been changed.

File: fe/fe-core/src/main/java/com/starrocks/service/FrontendServiceImpl.java (hunk 5 above)
Comment (ylcq) [reply fe/fe-core/src/main/java/com/starrocks/service/FrontendServiceImpl.java:542]: This is the previous code logic, which has not been changed. My understanding is to get values ​​like "PRI, UNI, DUP, AGG".

File: be/src/exec/vectorized/schema_scanner/schema_columns_scanner.cpp (hunk 6 above)
Comment (ylcq) [reply be/src/exec/vectorized/schema_scanner/schema_columns_scanner.cpp:205]: This is query optimization for "select xxx from information_schema.columns limit 10;", which only call describle_table interface once, so it is different from the previous logic of taking the dbname, so it is necessary to judge if else to get the dbname value

**Comment by kangkaisen** [comment]
run starrocks_fe_unittest

**Review by murphyatwork — APPROVED** [review]

**Review by stdpain — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Comment by kangkaisen** [comment]
for https://github.com/StarRocks/starrocks/issues/4028

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 0 / 57 (00.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|0|57|00.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 0 / 57 (00.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|0|57|00.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 0 / 57 (00.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|0|57|00.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 0 / 57 (00.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|0|57|00.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 0 / 57 (00.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|0|57|00.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 0 / 57 (00.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|0|57|00.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 0 / 57 (00.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|0|57|00.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 0 / 57 (00.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|0|57|00.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 0 / 57 (00.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|0|57|00.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 0 / 57 (00.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|0|57|00.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 0 / 57 (00.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|0|57|00.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 0 / 57 (00.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|0|57|00.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 0 / 57 (00.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|0|57|00.00%|

**Pull request merged**